### PR TITLE
Version1_Iteration_Plan_Week2

### PR DESCRIPTION
**Pull request by IonutCicio**
This is the first proposal for the implementation of the class "key".

### CODE REVIEW

File: KeyboardTeacher/Key.pde [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  void keyColor(color s, color f, color t) {
+    strokeColor = s;
+    fillColor = f;
+    textColor = t;
+  }
+
+  void dinamicColorChange(color a1, color a2, color a3, color b1, color b2, color b3, color c1, color c2, color c3, int tra) {
+    transparency = tra;
```
Comment (wbigger) [inline KeyboardTeacher/Key.pde:24]: If you want to assign a parameter to a variable of the class, a very common pattern is:
```this.transparency = trasparency;```
In this way you do not have to invent new exoteric variable names, and everything is plain and clear.

File: KeyboardTeacher/Key.pde [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    textSize = text.length() / keyWidth;
+  }
+
+  void keyColor(color s, color f, color t) {
+    strokeColor = s;
+    fillColor = f;
+    textColor = t;
+  }
+
+  void dinamicColorChange(color a1, color a2, color a3, color b1, color b2, color b3, color c1, color c2, color c3, int tra) {
```
Comment (wbigger) [inline KeyboardTeacher/Key.pde:23]: Give more descriptive names to parameters. What do a1/2/3 serve to? What about b? And c?

Comment (wbigger) [inline KeyboardTeacher/Key.pde:23]: Typo: dinamic -> dynamic

File: KeyboardTeacher/Key.pde [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    transparency = tra;
+    if (keyPressed) {
```
Comment (wbigger) [inline KeyboardTeacher/Key.pde:25]: Please add some comment to this function. Too many nested ifs, it is difficult to understand what is going on.

**Comment by wbigger** [comment]
You do not have absolutely to create a new class Key that is almost equal to Button. As you said, you could create a new parent class and extends both Button and Key, but I do not recommend it. "Less is more", and you should reduce the number of classes as much as you can.

To K.I.S.S. (Keep It Simple and Stupid), I would suggest to keep Button as it is, without any modification, and define the new Key class as a child. Something like this:

```
class Key extends Button {
  Key(int x, int y, int w, int h, String t) {
    // call the Button constructor
    super(x,y,w,h,t);
  }
  
  void dinamicColorChange(color a1, color a2, color a3, color b1, color b2, color b3, color c1, color c2, color c3, int trasparency) {
    // stuff here ...
  }
}
```

**Comment by IonutCicio** [comment]
It is true that I could define "Key" as a class child of "Button", but the issue with this implementation, is that "Key" would have "buttonClicked" and "mouseInside" In it. It's like you want a "Dog", but you get a "Cat" with that "Dog", in that way "Dog" can "miaow" . It would be more cleaner to create a class "Animal", and extend "Dog" and "Cat" from it.

**Comment by wbigger** [comment]
OK, good point about buttonClicked and mouseInside.

A couple of other hints:
- variables and methods names in a class should not mention the name of the class itself, to avoid inconsistencies when subclassing
- it is a good practice (if it is not too obvious or redundant) define the name of methods with a verb (remember that a method is an action).

I would suggest renaming variables and function with something like this:
```
boxWidth, boxHeight   --> width, height
boxColor()  --> setColors()
dinamicColorChange()  --> changeDynamicColors()
```

Look at [Processing reference](https://processing.org/reference/) for a good inspiration for naming.

**Pull request closed without merging**